tests/test_utils.py

```diff
@@ -0,0 +1,181 @@
+
+
+    # Apply function
+    result = preprocessing_model.upsample_month_end(
+        df,
+        datetime_col="DATETIME",
+        group_cols=["FUND_NAME", "SYMBOL"]
+    )
+
+    # Convert to dict for easy assertions
+    result_dict = {
+        row["DATETIME"].strftime("%Y-%m-%d") if hasattr(row["DATETIME"], "strftime") else row["DATETIME"]: row["PRICE"]
+        for row in result.to_dicts()
+    }
+    # 1. Original dates must be present
+    for date_str in ["2023-02-15", "2023-02-28", "2023-03-10", "2023-04-30"]:
+        assert date_str in result_dict
+
+    # 2. Expected generated month-end: 2023-03-31 (forward-fill from 2023-03-10)
+    assert "2023-03-31" in result_dict
+    assert result_dict["2023-03-31"] == 102
+
+    # 3. No overwriting of original values
+    assert result_dict["2023-03-10"] == 102
+    assert result_dict["2023-02-15"] == 95
+
+    # 4. Final row count = original rows + missing month-ends (1 added row)
+    assert result.shape[0] == 5
+
+@patch("app.etl.pre_process_data.read_file_as_string")
+@patch("app.etl.pre_process_data.execute_query")
+@patch("app.etl.pre_process_data.write_df_to_sqlite")
+def test_preprocess_reference_price_data_fix_date(
+    mock_write,
+    mock_execute,
+    mock_read_file,
+    mock_config,
+        sample_reference_df
+):
+    # Setup mock returns
+    mock_read_file.return_value = "SELECT * FROM equity_price"
+    mock_execute.return_value = sample_reference_df
+
+    # Create instance
+    preprocessor = PreprocessData(app_config=mock_config)
+    preprocessor.reference_data_query_file_name = "dummy.sql"
+
+    # Mock upsample method
+    preprocessor.upsample_month_end = MagicMock(return_value=sample_reference_df.with_columns(
+        pl.col("DATETIME").str.strptime(pl.Date, "%m/%d/%Y").cast(pl.Date)
+    ))
+
+    # Run method
+    preprocessor.preprocess_reference_price_data_fix_date()
+
+    # Assertions
+    mock_read_file.assert_called_once_with("dummy.sql")
+    mock_execute.assert_called_once()
+    mock_write.assert_called_once()
+
+    # Check if DATETIME was converted to YYYY-MM-DD format
+    written_df = mock_write.call_args[0][1]  # second arg is the DataFrame
+    assert "DATETIME" in written_df.columns
+    assert written_df["DATETIME"].dtype == pl.Utf8
+    assert all(len(str(date)) == 10 for date in written_df["DATETIME"])  # YYYY-MM-DD
+
+@patch("src.app.etl.generate_perf_report.GeneratePerfReport.generate_best_performing_fund_report")
+def test_compute_calls_generate_report(mock_generate, mock_config):
+    with patch.object(GeneratePerfReport, "generate_best_performing_fund_report") as mock_generate:
+        report = GeneratePerfReport(mock_config)
+        report._compute()
+        mock_generate.assert_called_once()
+
+@patch("pandas.DataFrame.to_excel")
+@patch("pandas.ExcelWriter")
+def test_export_filtered_joins_to_excel_polars(
+    mock_excel_writer, mock_to_excel, mock_config, recon_test_data
+):
```

Comment:
This is the first test I've seen that actually checks the output of the Excel writer!